### PR TITLE
Use tar-format to encapsulate data

### PR DESCRIPTION
**Pull request by TheMarex**
# Issue

This PR is another step towards implementing #4952. The goal is to make any `.osrm.*` file a `tar` file internally and address data by (file)name. As example the `.mldgr` file now looks like this:

```
/mld/multilevelgraph/node_array.meta
/mld/multilevelgraph/node_array
/mld/multilevelgraph/edge_array.meta
/mld/multilevelgraph/edge_array
/mld/multilevelgraph/node_to_edge_offset.meta
/mld/multilevelgraph/node_to_edge_offset
/mld/multilevelgraph/connectivity_checksum.meta
/mld/multilevelgraph/connectivity_checksum
```
Where every file contains raw binary data for data structures in OSRM that we can deserialize or read directly to memory. The `.meta` files contain additional information for setting up the data layout, e.g. the number of values for serialized `std::vector`, or the fingerprint.

## Tasklist

 - [x] add tests (see [testing documentation](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/testing.md)
 - [x] review
 - [x] adjust for comments

### CODE REVIEW

File: include/storage/serialization.hpp [hunk 1]
```diff
@@ -29,160 +33,190 @@ namespace serialization
...
-        writer.WriteOne<unsigned char>(packBits(data, CHAR_BIT * index, CHAR_BIT));
-    }
-    if (count > index)
-        writer.WriteOne<unsigned char>(packBits(data, index, count - index));
+
+    constexpr std::uint64_t WORD_BITS = CHAR_BIT * sizeof(std::uint64_t);
+
+    // FIXME on old boost version the function_input_iterator does not work with lambdas
+    // so we need to wrap it in a function here.
+    const std::function<std::uint64_t()> encode_function = [&]() -> char {
```
Comment (oxidase) [inline include/storage/serialization.hpp]: s/-> char/-> std::uint64_t/

File: include/storage/serialization.hpp [hunk 2]
```diff
@@ -29,160 +33,190 @@ namespace serialization
...
+        auto write_size = std::min<std::size_t>(count - index, WORD_BITS);
+        auto packed = packBits<VectorT, std::uint64_t>(data, index, write_size);
+        index += WORD_BITS;
+        return packed;
+    };
+
+    std::uint64_t number_of_blocks = std::ceil((double)count / WORD_BITS);
```
Comment (oxidase) [inline include/storage/serialization.hpp]: `(count + WORD_BITS - 1) / WORD_BITS` would be better to avoid floating-point arithmetic

Comment (TheMarex) [reply include/storage/serialization.hpp]: Ah nice trick. I know a few places were this would come in useful.

**Comment by oxidase** [comment]
@TheMarex i pushed fixes for msvc to trigger CI builds ("fixes" are reducing c++ syntactic sugar for msvc). Feel free to kick-out or squash the last commit.

**Comment by TheMarex** [comment]
> unknown location(0): fatal error: in "route/test_manual_setting_of_annotations_property": class osrm::util::exception: File "C:/projects/osrm/test/data/ch/monaco.osrm.fileIndex" mapping failed: failed opening file: The process cannot access the file because it is being used by another process.

Some more problems with windows. I suspect this is because I had to change the r-tree to allow writing to an mmap files. Maybe I can work around this.

**Comment by TheMarex** [comment]
Wuhu finally! It compiles. :rocket:

**Comment by TheMarex** [comment]
@mloskot you are right, this doesn't meet the single-file requirement yet.

However, since we now use named identifiers for data, `osrm-datastore` does not really need to care anymore in which files the data is stored, we could figure that out automatically. That would unlock just passing a single file to `osrm-datastore` that is just a bunch of merged tar files. I captured this in https://github.com/Project-OSRM/osrm-backend/issues/4978 for further discussion.

**Comment by mloskot** [comment]
@TheMarex It makes sense, thanks for the clarification.

**Pull request merged**